File: test/samples/8-queens.ml
type int_list = Nil | Cons of int * int_list

let rec safe qs col i = match qs with
   Nil -> true
 | Cons (h, t) ->
      if h = col || h - i = col || h + i = col then false
      else safe t col (i + 1);;

let rec solve n row p =
   if row = n then p else
   let rec loop col =
      if col = n then Nil
      else if safe p col 1
         then
            match solve n (row + 1) (Cons (col, p)) with
               Nil -> loop (col + 1)
             | s -> s
         else loop (col + 1)
   in loop 0;;

let exec n = solve n 0 Nil in
exec

File: test/samples/8-queens.js
const safe = (queens, col) => {
   for (let i = 1; i <= queens.length; i++) {
      const qcol = queens[queens.length - i]
      if (col == qcol || col + i == qcol || col - i == qcol) return false
   }
   return true
}

const solve = (n, queens) => {
   if (queens.length == n) return queens
   for (let col = 0; col < n; col++) {
      if (safe(queens, col)) {
         const solution = solve(n, [...queens, col])
         if (solution) return solution
      }
   }
   return null
}

console.log(solve(10, []).reverse())
